neo4j/driver_with_context.go

```diff
@@ -315,3 +369,144 @@ func (d *driverWithContext) Close(ctx context.Context) error {
+		if d.defaultManagedSessionBookmarkManager == nil { // this allows tests to init the field themselves
+			d.defaultManagedSessionBookmarkManager = NewBookmarkManager(BookmarkManagerConfig{})
+		}
+	})
+	return d.defaultManagedSessionBookmarkManager
+}
+
+func (d *driverWithContext) executeQueryCallback(ctx context.Context, query string, parameters map[string]any) ManagedTransactionWork {
+	return func(tx ManagedTransaction) (any, error) {
+		cursor, err := tx.Run(ctx, query, parameters)
+		if err != nil {
+			return nil, err
+		}
+		keys, err := cursor.Keys()
+		if err != nil {
+			return nil, err
+		}
+		records, err := cursor.Collect(ctx)
+		if err != nil {
+			return nil, err
+		}
+		summary, err := cursor.Consume(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return &EagerResult{
+			Keys:    keys,
+			Records: records,
+			Summary: summary,
+		}, nil
+	}
+}
+
+type ExecuteQueryConfigurationOption func(*ExecuteQueryConfiguration)
+
+func WithReadersRouting() ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.Routing = Readers
+	}
+}
+
+func WithWritersRouting() ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.Routing = Writers
+	}
+}
+
+func WithImpersonatedUser(user string) ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.ImpersonatedUser = user
+	}
+}
+
+func WithDatabase(db string) ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.Database = db
+	}
+}
+
+func WithBookmarkManager(bookmarkManager BookmarkManager) ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.BookmarkManager = bookmarkManager
+	}
+}
+
+func WithoutBookmarkManager() ExecuteQueryConfigurationOption {
+	return func(configuration *ExecuteQueryConfiguration) {
+		configuration.BookmarkManager = nil
+	}
+}
+
+type ExecuteQueryConfiguration struct {
+	Routing          RoutingControl
+	ImpersonatedUser string
+	Database         string
+	BookmarkManager  BookmarkManager
+}
+
+type RoutingControl int
+
+const (
+	Writers RoutingControl = iota
+	Readers
+)
+
+func (c *ExecuteQueryConfiguration) toSessionConfig() SessionConfig {
+	return SessionConfig{
+		ImpersonatedUser: c.ImpersonatedUser,
+		DatabaseName:     c.Database,
+		BookmarkManager:  c.BookmarkManager,
+	}
+}
+
+type transactionFunction func(context.Context, ManagedTransactionWork, ...func(*TransactionConfig)) (any, error)
+
+func (c *ExecuteQueryConfiguration) selectTxFunctionApi(session SessionWithContext) (transactionFunction, error) {
+	switch c.Routing {
+	case Readers:
+		return session.ExecuteRead, nil
+	case Writers:
+		return session.ExecuteWrite, nil
+	}
+	return nil, fmt.Errorf("unsupported routing control, got: %d", c.Routing)
+}
+
+type EagerResult struct {
+	Keys    []string
+	Records []*Record
```

Comment:
Out of curiosity: why are records pointers being passed around? Is that for a lower memory and runtime profile?

Reply:
I think for slices, it does not matter that much. I just reused the existing type of `ResultWithContext#Collect`.
In the general, passing a pointer of struct avoid copying the said struct instance.